backend-tests/tests/test_deployments.py

```diff
@@ -502,3 +583,315 @@ def compare_phases_json(expected, response):
+                user_token).call('GET', os.path.join('/deployments', deployment_id))
+            phases = resp.json()["phases"]
+            for p in range(i+1):
+                assert phases[p]["device_count"] == batch_sizes[p]
+            for p in range(i+1,len(phases)):
+                assert phases[p]["device_count"] == 0
+
+        # Finally confirm that deployment is finished
+        assert resp.json()["status"] == "finished"
+
+    def test_phased_regular_deployment(self, setup_devices_and_management):
+        """
+        Phased equivalent of a regular deployment.
+        """
+        user, tenant, utoken, devs = setup_devices_and_management()
+
+        deployment_req = {
+            "name"          : "phased-regular-deployment",
+            "artifact_name" : "deployments-phase-testing",
+            "devices"       : [dev.id for dev in devs],
+            "phases"        : [{}]
+        }
+        self.try_phased_updates(deployment_req, devs, utoken)
+
+
+    def test_delayed_deployment(self, setup_devices_and_management):
+        """
+        Uses a single phase with a delayed start-time.
+        """
+        user, tenant, utoken, devs = setup_devices_and_management()
+
+        deployment_req = {
+            "name"          : "phased-delayed-deployment",
+            "artifact_name" : "deployments-phase-testing",
+            "devices"       : [dev.id for dev in devs],
+            "phases"        : [
+                {"start_ts" : (datetime.utcnow() + timedelta(
+                               seconds=15)).strftime("%Y-%m-%dT%H:%M:%SZ")}
+            ]
+        }
+        self.try_phased_updates(deployment_req, devs, utoken)
+
+
+    def test_two_phases_full_spec(self, setup_devices_and_management):
+        """
+        Two phases, with batch_size and start_ts specified for both phases.
+        """
+        user, tenant, utoken, devs = setup_devices_and_management()
+        deployment_req = {
+            "name"    : "two-fully-spec-deployments",
+            "artifact_name" : "deployments-phase-testing",
+            "devices" : [dev.id for dev in devs],
+            "phases"        : [
+                {
+                    "batch_size" : 10,
+                    "start_ts"   : (datetime.utcnow() + timedelta(
+                                    seconds=15)).strftime("%Y-%m-%dT%H:%M:%SZ")
+                },
+                {
+                    "start_ts"   : (datetime.utcnow() + timedelta(
+                                    seconds=30)).strftime("%Y-%m-%dT%H:%M:%SZ"),
+                    "batch_size" : 90
+                }
+            ]
+        }
+        self.try_phased_updates(deployment_req, devs, utoken)
+
+
+    def test_three_phased_deployments(self, setup_devices_and_management):
+        """
+        Three phases; with no start_ts in first and no batch_size in third.
+        """
+        user, tenant, utoken, devs = setup_devices_and_management(nr_devices=101)
```

Comment:
I almost said you're calling the fixture twice - once in the argument, and once here.

I can see though that you're doing a clever trick, and the fixture simply returns a func. that's clever, and I'm not sure what to do with it. so do I understand correctly that:
- by having the fixture as arg, we get clean mongo + a func we can call - but no init data yet
- this call is what actually does the init?

I assume it's a way of parametrizing a fixture. pytest does provide fixture parametrization though - it's a slightly advanced pattern with the `indirect` keyword. by advanced I mean - unintuitive and confusing.

my suggestion would be to simply get rid of the fixture, and call it just as a func. so:
- `setup_devices_and_management` remains an ordinary func - returns result, not another func
- the fixture decorator is removed
- each test accepts the `clean_mongo` fixture instead.

as a result stuff is simple. you call setup_ as usual in every test, we don't fiddle with fixture parametrization, no second-order functions fly around the codebase, each test gets a clean db as it should.

Reply:
I agree. My mind was fixed on using this setup function as a fixture that I came up with this workaround. Explicitly calling the function is much cleaner.